src/DynamoCore/Visualization/RenderPackageCache.cs

```diff
@@ -0,0 +1,128 @@
+
+namespace Dynamo.Visualization
+{
+    /// <summary>
+    /// This class controls the collection and distribution of render packages
+    /// </summary>
+    public class RenderPackageCache
+    {
+        private List<IRenderPackage> packages;
+        private Dictionary<Guid, RenderPackageCache> portMap;
+
+        private void AddPort(IRenderPackage package, Guid outputPortId)
+        {
+            if (portMap == null)
+            {
+                portMap = new Dictionary<Guid, RenderPackageCache>();
+            }
+
+            if (!portMap.ContainsKey(outputPortId))
+            {
+                portMap[outputPortId] = new RenderPackageCache();
+            }
+
+            portMap[outputPortId].Add(package);
+        }
+
+        /// <summary>
+        /// Create an empty RenderPackageCache object
+        /// </summary>
+        public RenderPackageCache()
+        {
+            packages = new List<IRenderPackage>();
+            portMap = null;
+        }
+
+        /// <summary>
+        /// Create a RenderPackageCache object containing the given render packages
+        /// </summary>
+        /// <param name="otherPackages">The packages to fill the new cache object.</param>
+        public RenderPackageCache(IEnumerable<IRenderPackage> otherPackages)
+        : this()
+        {
+            packages.AddRange(otherPackages);
+        }
+
+        /// <summary>
+        /// Get the render packages in this cache
+        /// </summary>
+        public IEnumerable<IRenderPackage> Packages
+        {
+            get
+            {
+                return packages;
+            }
+        }
+
+        /// <summary>
+        /// Get the RenderPackageCache object for the given port ID
+        /// </summary>
+        /// <param name="portId">The port ID used to get the sub-cache.</param>
+        public RenderPackageCache GetPortPackages(Guid portId)
+        {
+            if (portMap == null)
+                return null;
+
+            RenderPackageCache portPackages;
+            if (!portMap.TryGetValue(portId, out portPackages))
+                return null;
+
+            if (portPackages == null)
+                return null;
+
+            return portPackages;
+        }
+
+        /// <summary>
+        /// Returns true if the cache is empty
+        /// </summary>
+        public bool IsEmpty()
+        {
+            return packages.Count == 0;
+        }
+
+        /// <summary>
+        /// Concatenates the other cache into this cache
+        /// </summary>
+        /// <param name="other">The cache to add to this cache.</param>
+        public void Add(RenderPackageCache other)
+        {
+            packages.AddRange(other.packages);
+
+            if (other.portMap == null)
+                return;
+
+            foreach (var port in other.portMap)
+            {
+                foreach(var item in port.Value.packages)
+                {
+                    AddPort(item, port.Key);
```

Comment:
what happens if you have duplicate ports here, it will replace them?

Reply:
maybe we should assert that this is not the case?

Reply:
This adds the render package to a list of render packages indexed by key (and lazily created only when needed). So, duplicate ports will cause the render packages to be concatenated into a single list.

Reply:
I think that there should not be duplicate ports here, although this may not be true. I can try adding an assert here.